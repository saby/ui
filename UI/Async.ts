/**
 * Библиотека методов для асинхронной загрузки контролов.
 * @library UI/Async
 * @public
 * @author Мустафин Л.И.
 */

export { default as Async } from './_async/Async';
export { lazy, importer, ILazyProps } from './_async/lazy';
