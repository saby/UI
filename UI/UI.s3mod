<?xml version="1.0" encoding="UTF-8" ?>
<ui_module id="f27abe12-19df-4c9f-b886-634bd20a0011" name="UI" package="Платформа" responsible="Шипин А.А." required="1" version="1.00">

  <description>Библиотека, позволяющая собрать "каркас" веб-приложения; содержит набор правил, техник и решений, допустимых при разработке прикладных решений.</description>

  <depends>
    <module id="0df0fce3-53ba-47f8-987d-edf1fd078295" name="WS.Core"/>
    <module id="10efeac7-0f02-4c73-9474-0d1d0db077b6" name="View"/>
    <module id="636177e5-4d75-44fa-9173-977dc90ff111" name="Env"/>
    <module id="0f5d26db-dd61-4957-9375-8861ad66f219" name="Vdom"/>
    <module id="63e1a7e5-4d75-42fa-91a3-977dc90ff1e1" name="SbisEnv"/>
    <module id="323fd54f-1fab-42a6-8aea-6726d6392d53" name="Application"/>
    <module id="832cda0a-e19c-4ffc-a1d1-47edbb6b7501" name="I18n"/>
  </depends>

  <load_after/>

  <parameters/>

  <resources/>

  <preload/>

</ui_module>
