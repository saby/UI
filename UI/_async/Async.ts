/// <amd-module name='UI/Base:Async' />
import * as ModulesLoader from 'WasabyLoader/ModulesLoader';
import * as Library from 'WasabyLoader/Library';
import { IoC, constants } from 'Env/Env';
import { descriptor } from 'Types/entity';
import { default as Control, IControlOptions, TemplateFunction } from 'UI/_base/Control';
import { headDataStore } from 'UI/_base/HeadData';
import template = require('wml!UI/_async/Async');

function generateErrorMsg(templateName: string, msg?: string): string {
   const tTemplate = `Ошибка загрузки контрола "${templateName}"`;
   const tHint = 'Возможны следующие причины:\n\t \
                  • Ошибка в самом контроле\n\t \
                  • Долго отвечал БЛ метод в _beforeUpdate\n\t \
                  • Контрола не существует';
   return !msg ? `${tTemplate}\n${tHint}` : `${tTemplate}: ${msg}`;
}

export type TAsyncStateReceived = boolean | string;

export interface IAsyncOptions extends IControlOptions {
   templateName: string;
   templateOptions: IControlOptions;
}

const SUCCESS_BUILDED = 's';

/**
 * Контейнер для асинхронной загрузки контролов.
 * Подробное описание и примеры вы можете найти <a href='/doc/platform/developmentapl/interface-development/pattern-and-practice/async-load/'>здесь</a>.
 *
 * @class UI/Base:Async
 * @extends Core/Control
 *
 * @private
 * @author Санников К.А.
 */
export default abstract class Async extends Control<IAsyncOptions, TAsyncStateReceived> {
   protected _template: TemplateFunction = template;
   protected currentTemplateName: string;
   protected optionsForComponent: Record<string, unknown> = {};
   /**
    * Флаг для того, чтобы избежать повторной загрузки шаблона, при изменении опций до окончания асинхронной загрузки
    */
   private asyncLoading: boolean = false;
   /**
    * Флаг, о том, что произошла ошибка при загрузке модуля - чтобы не было циклической попытки загрузки
    */
   private loadingErrorOccurred: boolean = false;
   protected error: TAsyncStateReceived | void;
   protected userErrorMessage: string | void;
   protected defaultErrorMessage: string = 'У СБИС возникла проблема';

   protected _beforeMount(options: IAsyncOptions, _: unknown, receivedState: TAsyncStateReceived): Promise<TAsyncStateReceived> {
      if (!options.templateName) {
         this.error = 'В модуль Async передали не корректное имя шаблона (templateName=undefined|null|empty)';
         IoC.resolve('ILogger').error(this.error);
         return Promise.resolve(this.error);
      }

      if (receivedState && receivedState !== SUCCESS_BUILDED) {
         IoC.resolve('ILogger').error(receivedState);
      }

      if (constants.isBrowserPlatform && (!ModulesLoader.isLoaded(options.templateName) ||
         this._isCompat() || !receivedState)) {
         return this._loadContentAsync(options.templateName, options.templateOptions);
      }

      this.error = this._loadContentSync(options.templateName, options.templateOptions);
      if (this.error) {
         this.userErrorMessage = this.defaultErrorMessage;
         return Promise.resolve(this.error);
      }

      return Promise.resolve(SUCCESS_BUILDED);
   }

   /**
    * Если можем подставить данные при изменении синхронно, то делаем это до обновления
    * @param {*} opts
    */
   protected _beforeUpdate(opts: IAsyncOptions): void {
      if (this.asyncLoading) {
         return;
      }

      if (opts.templateName === this.currentTemplateName) {
         // поменялись только опции шаблона
         this._insertComponent(this.optionsForComponent.resolvedTemplate,
            opts.templateOptions,
            opts.templateName);
         return;
      }

      if (ModulesLoader.isLoaded(opts.templateName)) {
         this._loadContentSync(opts.templateName, opts.templateOptions);
      }
   }

   /**
    * Если до обновления не загрузили синхронно, то пробуем загрузить асинхронно
    */
   protected _afterUpdate(): void {
      if (this.asyncLoading) {
         return;
      }
      if (this.loadingErrorOccurred) {
         this.loadingErrorOccurred = false;
         return;
      }
      if (this.currentTemplateName === this._options.templateName) {
         return;
      }

      this._loadContentAsync(this._options.templateName, this._options.templateOptions).then(() => {
         this._forceUpdate();
      });
   }

   protected _loadContentSync(name: string, options: IControlOptions): TAsyncStateReceived {
      const loaded = this._loadSync(name);
      if (loaded === null) {
         return generateErrorMsg(name);
      }

      this._insertComponent(loaded, options, name);
      this._pushDepToHeadData(Library.parse(name).name);
      return false;
   }

   protected _loadSync<T = unknown>(name: string): T {
      try {
         const loaded = ModulesLoader.loadSync<T>(name);
         if (loaded) {
            return loaded;
         }
      } catch (err) {
         IoC.resolve('ILogger').error(`Couldn't load module "${name}"`, err);
      }
      return null;
   }

   protected _loadContentAsync(name: string, options: IControlOptions): Promise<TAsyncStateReceived> {
      // Need this flag to prevent setting new options for content
      // that wasn't loaded yet
      this.asyncLoading = true;
      this.loadingErrorOccurred = false;

      return this._loadAsync(name).then<TAsyncStateReceived, TAsyncStateReceived>((loaded) => {
         this.asyncLoading = false;
         if (!loaded) {
            this.loadingErrorOccurred = true;
            this.error = generateErrorMsg(name);
            IoC.resolve('ILogger').warn(this.error);
            this.userErrorMessage = this.defaultErrorMessage;
            return this.error;
         }

         this._insertComponent(loaded, options, name);
         return true;
      }, (err) => {
         this.asyncLoading = false;
         this.loadingErrorOccurred = true;
         this.error = generateErrorMsg(name);
         this.userErrorMessage = err.message;
         return err;
      });
   }

   protected _loadAsync(name: string): Promise<unknown> {
      return ModulesLoader.loadAsync(name).catch((error) => {
         IoC.resolve('ILogger').error(`Couldn't load module "${name}"`, error);
         ModulesLoader.unloadSync(name);
         throw new Error(this.defaultErrorMessage);
      });
   }

   protected _pushDepToHeadData(dep: string): void {
      if (constants.isBrowserPlatform) {
         return;
      }

      try {
         headDataStore.read('pushDepComponent')(dep, true);
      } catch (e) {
         IoC.resolve('ILogger').warn('You\'re trying to use Async without Controls/Application. Link to ' +
            dep +
            ' won\'t be added to server-side generated markup. ' + e);
      }
   }

   protected _insertComponent(tpl: unknown, opts: IControlOptions, templateName: string): void {
      this.error = '';
      this.currentTemplateName = templateName;
      this.optionsForComponent = {};
      for (const key in opts) {
         if (opts.hasOwnProperty(key)) {
            this.optionsForComponent[key] = opts[key];
         }
      }

      if (tpl && tpl['__esModule']) {
         this.optionsForComponent.resolvedTemplate = tpl['default'];
         return;
      }
      this.optionsForComponent.resolvedTemplate = tpl;
   }

   protected _isCompat(): boolean {
      return constants.compat;
   }

   static getOptionTypes(): Record<string, unknown> {
      return {
         templateName: descriptor(String).required()
      };
   }
}

/**
 * @name UI/Base:Async#content
 * @cfg {String} Содержимое контейнера.
 */

/**
 * @name UI/Base:Async#templateName
 * @cfg {String} Имя асинхронно загружаемого контрола.
 * Можно использовать только {@link /doc/platform/developmentapl/interface-development/pattern-and-practice/javascript-libraries/#_2 публичные пути библиотеки}.
 */

/**
 * @name UI/Base:Async#templateOptions
 * @cfg {Object} Параметры содержимого контейнера Async.
 */
