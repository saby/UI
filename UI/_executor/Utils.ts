/// <amd-module name="UI/_executor/Utils" />
/* tslint:disable */

/**
 * @author Тэн В.А.
 */

import * as Class from './_Utils/Class';
import * as Common from './_Utils/Common';
import * as ConfigResolver from './_Utils/ConfigResolver';
import * as OptionsResolver from './_Utils/OptionsResolver';
import * as RequireHelper from './_Utils/RequireHelper';
import * as Vdom from './_Utils/Vdom';
import VoidTags from './_Utils/VoidTags';

export {
   Class,
   Common,
   ConfigResolver,
   OptionsResolver,
   RequireHelper,
   Vdom,
   VoidTags
};
