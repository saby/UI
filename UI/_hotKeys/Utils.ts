/**
 * Вспомогательная библиотека для работы горячих клавиш
 * @class UI/_utils/HotKeys
 * @author Тэн В.А.
 */

export { keyDownHandler as dispatcherHandler, ISyntheticEvent } from './Utils/dispatcher';
