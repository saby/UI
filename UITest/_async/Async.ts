import { Async, TAsyncStateReceived, IAsyncOptions } from 'UICore/Async';

/**
 * Реализация класса UICore/Async:Async для тестов
 */
export default class AsyncTest extends Async {
    _beforeMount(options: IAsyncOptions): void {
        super._beforeMount(options);
    }

    _componentDidMount() {
        super._componentDidMount();
    }

    _beforeUpdate(opts: IAsyncOptions) {
        super._beforeUpdate(opts);
    }

    _afterUpdate() {
        super._afterUpdate();
    }

    getError(): TAsyncStateReceived | void {
        return this.error;
    }

    getCurrentTemplateName(): string {
        return this.currentTemplateName;
    }

    getOptionsForComponent(): Record<string, unknown> {
        return this.optionsForComponent;
    }
}
