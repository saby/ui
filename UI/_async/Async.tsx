import { ReactNode, Component, JSXElementConstructor } from 'react';
import { isLoaded, loadSync, loadAsync, unloadSync } from 'WasabyLoader/ModulesLoader';
import { parse } from 'WasabyLoader/Library';
import { addPageDeps } from 'UI/Deps';
import { Logger } from 'UI/Utils';
import { CommonUtils } from 'UI/Executor';

interface IAsyncProps {
    componentName: string;
    componentProps?: Record<string, unknown>;
    onComponentLoad?: () => {};
    children?: ReactNode;
    className?: string;
}

interface IAsyncState {
    currentComponentName?: string;
    component?: JSXElementConstructor<any>;
    error?: boolean;
    userErrorMessage?: string;
}

type TContentConstructor = JSXElementConstructor<any> | undefined;

const isServerSide = typeof window === 'undefined';

/**
 * Компонент для асинхронной загрузки контролов
 * @private
 */
export default class Async extends Component<IAsyncProps, IAsyncState> {
    /**
     * Флаг для того, чтобы избежать повторной загрузки шаблона, при изменении опций до окончания асинхронной загрузки
     */
    protected asyncLoading: boolean = false;

    /**
     * Флаг, о том, что произошла ошибка при загрузке модуля - чтобы не было циклической попытки загрузки
     */
    private loadingErrorOccurred: boolean = false;

    private defaultErrorMessage: string = 'У СБИС возникла проблема';

    /**
     * Флаг чтобы понимать, что был загружен контрол и вставлен на страницу -
     * т.к. после построения нужно будет вызвать коллбек onComponentLoad
     */
    private needNotifyOnLoad: boolean = false;

    constructor(props: IAsyncProps) {
        super(props);

        if (!props.componentName) {
            const error =
                'В модуль Async передали не корректное имя шаблона (componentName=undefined|null|empty)';
            Logger.error(error);
            this.state = {};
            return;
        }

        let state: IAsyncState = {};
        if (isServerSide || isLoaded(props.componentName)) {
            const component = this._loadContentSync(props.componentName);
            state = {
                component,
                currentComponentName: props.componentName,
                error: !component,
            };
        }

        if (state.error) {
            state.userErrorMessage = this.defaultErrorMessage;
        }

        this.state = state;
    }

    componentDidMount(): void {
        if (this.state.component) {
            this._callOnComponentLoad();
            return;
        }

        this._loadContentAsync(this.props.componentName);
    }

    componentDidUpdate(): void {
        if (this.asyncLoading) {
            return;
        }

        if (this.loadingErrorOccurred) {
            this.loadingErrorOccurred = false;
            return;
        }

        if (this.state.currentComponentName === this.props.componentName) {
            this._callOnComponentLoad();
            return;
        }

        if (isLoaded(this.props.componentName)) {
            const component = this._loadContentSync(this.props.componentName);
            this.setState({
                currentComponentName: this.props.componentName,
                component,
                error: false,
            });
            return;
        }

        this._loadContentAsync(this.props.componentName);
    }

    render() {
        if (this.state.error || !this.state.component) {
            return <div className={this.props.className}>{this.state.userErrorMessage}</div>;
        }

        if (this.state.currentComponentName === this.props.componentName) {
            this._callOnComponentLoad();
        }

        if (!this.props.children) {
            return (
                <this.state.component
                    className={this.props.className}
                    {...this.props.componentProps}
                />
            );
        }

        return (
            <this.state.component className={this.props.className} {...this.props.componentProps}>
                {this.props.children}
            </this.state.component>
        );
    }

    private _callOnComponentLoad(): void {
        if (this.needNotifyOnLoad && !this.state.error && !this.asyncLoading) {
            this.needNotifyOnLoad = false;
            this.props.onComponentLoad?.();
        }
    }

    private _loadContentSync(componentName: string): TContentConstructor {
        const loaded = this._loadSync(componentName);
        if (!loaded) {
            return;
        }

        this.needNotifyOnLoad = true;
        addPageDeps([parse(componentName).name]);
        return getComponent(loaded);
    }

    private _loadSync<T = TContentConstructor>(name: string): T | undefined {
        try {
            const loaded = loadSync<T>(name);
            if (loaded) {
                return loaded;
            }
        } catch (err) {
            Logger.error(`Couldn't load module "${name}"`, err);
        }
    }

    private _loadContentAsync(componentName: string): Promise<void> {
        this.asyncLoading = true;
        this.loadingErrorOccurred = false;

        return this._loadAsync(componentName).then(
            (loaded) => {
                this.asyncLoading = false;
                if (!loaded) {
                    this.loadingErrorOccurred = true;
                    const error = generateErrorMsg(componentName);
                    Logger.warn(error);
                    this.setState({ error: true, userErrorMessage: this.defaultErrorMessage });
                    return;
                }

                this.needNotifyOnLoad = true;
                this.setState({
                    currentComponentName: componentName,
                    component: getComponent(loaded),
                    error: false,
                });
            },
            (err) => {
                this.asyncLoading = false;
                this.loadingErrorOccurred = true;
                this.setState({ error: true, userErrorMessage: err.message });
            }
        );
    }

    private _loadAsync(name: string): Promise<TContentConstructor> {
        return loadAsync<TContentConstructor>(name).catch((error) => {
            Logger.error(`Couldn't load module "${name}"`, error);
            unloadSync(name);
            throw new Error(this.defaultErrorMessage);
        });
    }

    static displayName: string = 'UI/Async:Async';
}

function generateErrorMsg(componentName: string, msg?: string): string {
    const tTemplate = `Ошибка загрузки контрола "${componentName}"`;
    const tHint =
        'Возможны следующие причины:\n\t \
                  • Ошибка в самом контроле\n\t \
                  • Долго отвечал БЛ метод в _beforeUpdate\n\t \
                  • Контрола не существует';
    return !msg ? `${tTemplate}\n${tHint}` : `${tTemplate}: ${msg}`;
}

function getComponent(component: TContentConstructor): TContentConstructor {
    if (component && CommonUtils.isDefaultExport(component)) {
        // @ts-ignore
        return component.default;
    }
    return component;
}
