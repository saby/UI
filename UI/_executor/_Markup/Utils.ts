/// <amd-module name="UI/_executor/_Markup/Utils" />
/* tslint:disable */

import * as Attr from '../_Expressions/Attr';
import * as Decorate from '../_Expressions/Decorate';
import { _FocusAttrs } from 'UI/Focus';
import {
   GeneratorEmptyObject, GeneratorError, GeneratorFn,
   GeneratorStringArray,
   IBaseAttrs, IGeneratorConfig,
   IGeneratorDefCollection, IStringTemplateResolverIncludedTemplates,
   TAttributes,
   TObject
} from './IGeneratorType';
import * as Common from '../_Utils/Common';
import voidElements from '../_Utils/VoidTags';
// @ts-ignore
import { Number } from 'UI/Utils';
import { INodeAttribute } from './IGeneratorType';
import {IAttributes} from '../_Expressions/Attr';

/**
 * @author Тэн В.А.
 */

interface IControlData {
   name?: unknown;
}

// joinAttrs = Attr.joinAttrs;

/**
 * Понимаем асинхронная ветка или нет
 * @param entity
 * @returns {unknown}
 */
export function isInstOfPromise(entity: Promise<any>): unknown {
   return entity && entity.then;
}

/**
 * Создаем строку с тегом для повторного выполнения
 * _beforeMount на клиенте и обработки ошибок
 * @param inst
 * @param createTag
 * @returns {string}
 */
export function asyncRenderErrorTag(inst: TObject, createTag?: Function): string {
   let decoratorObject = {};
   let options;
   if (inst && inst._options) {
      options = inst._options;
      decoratorObject = Decorate.createRootDecoratorObject(
         options.__$config,
         true,
         options['data-component'],
         {}
      );
   }
   const createTagFn = createTag ? createTag : createTagDefault;
   return createTagFn('div', { attributes: decoratorObject }, []);
}

/**
 *
 * @param markup
 * @returns {string}
 */
export function invisibleNodeCompat(markup: string): string {
   return markup && markup.indexOf && markup.indexOf('<invisible-node') === 0 ?
      markup.replace(/invisible-node/g, 'div') : markup;
}

/**
 *
 * @param tag
 * @param attrs Собственные атрибуты
 * @param children
 * @param attrToDecorate
 * @param defCollection
 * @param control
 * @returns {string}
 */
export function createTagDefault(tag: string,
                                 attrs: IBaseAttrs | {attributes: unknown},
                                 children: GeneratorStringArray,
                                 attrToDecorate?: TAttributes,
                                 defCollection?: IGeneratorDefCollection,
                                 control?: GeneratorEmptyObject): string {
   if (!attrToDecorate) {
      attrToDecorate = {};
   }
   if (!attrs) {
      attrs = {attributes: {}};
   }

   let mergedAttrs = Attr.processMergeAttributes(
       attrToDecorate.attributes as IAttributes,
       attrs.attributes as IAttributes,
       true
   );

   _FocusAttrs.prepareTabindex(mergedAttrs);
   // remove focus attributes from object
   if (Common.isCompat()) {
      // ! не вырезаем фокусные атрибуты, для совместимости. чтобы старые компоненты могли работать в новом окружении
      // textMarkupGenerator.cutFocusAttributes(mergedAttrs);
   } else {
      cutFocusAttributes(mergedAttrs);
   }
   const mergedAttrsStr = mergedAttrs
      ? decorateAttrs(mergedAttrs, {})
      : '';
   // tslint:disable-next-line:no-bitwise
   if (~voidElements.indexOf(tag)) {
      return '<' + tag + mergedAttrsStr + ' />';
   }
   return '<' + tag + mergedAttrsStr + '>' + joinElements(children, undefined, defCollection) + '</' + tag + '>';
}

/**
 *
 * @param attr1
 * @param attr2
 * @returns {string}
 */
export function decorateAttrs(attr1: TAttributes, attr2: TAttributes): string {
   function wrapUndef(value: string): string {
      if (value === undefined || value === null) {
         return '';
      } else {
         return value;
      }
   }

   const attrToStr = (attrs: Array<string>): string => {
      let str = '';
      for (const attr in attrs) {
         if (attrs.hasOwnProperty(attr)) {
            str += (wrapUndef(attrs[attr]) !== '' ? ' ' + (attr + '="' + attrs[attr] + '"') : '');
         }
      }
      return str;
   };
   return attrToStr(Attr.joinAttrs(attr1, attr2));
}

const focusAttrs = [
   'ws-creates-context',
   'ws-delegates-tabfocus',
   'ws-tab-cycling',
   'ws-no-focus',
   'attr:ws-creates-context',
   'attr:ws-delegates-tabfocus',
   'attr:ws-tab-cycling',
   'attr:ws-no-focus'
];

/**
 * Скрывает атрибуты необходимые для работы системы фокусов
 * @param attributes
 * @param fn
 * @param node
 * @return {object}
 */
export function cutFocusAttributes(attributes: TAttributes, fn?: Function, node?: HTMLElement): void {
   focusAttrs.forEach((focusAttr: string): void => {
      if (attributes.hasOwnProperty(focusAttr)) {
         fn && fn(focusAttr, attributes[focusAttr]);
         delete attributes[focusAttr];
         if (node) {
            node.removeAttribute(focusAttr);
         }
      }
   });
}

/**
 * Рекуриснове объединение элементов, если пришел массив из partial
 * @param elements
 * @param key
 * @param defCollection
 * @return {Array<object | string> | string | Error}
 */
export function joinElements(elements: Array<unknown>,
                             key?: string,
                             defCollection?: IGeneratorDefCollection): GeneratorStringArray | GeneratorError {
   if (Array.isArray(elements)) {
      let res = '';
      elements.forEach(function joinOneElement(element) {
         let id;
         if (Array.isArray(element)) {
            element = joinElements(element, undefined, defCollection);
         }
         if (element && isInstOfPromise(element)) {
            id = Number.randomId('def-');
            if (!defCollection.def) {
               defCollection.def = [];
            }
            defCollection.def.push(element);
            element = '[' + id + ']';
            defCollection.id.push(element);
         }
         res += (element || '');
      });

      return res;
   } else {
      throw new Error('joinElements: elements is not array');
   }
}

export function resolveControlName<TOptions extends IControlData>(controlData: TOptions,
                                                                  attributes: TAttributes | INodeAttribute): TAttributes | INodeAttribute {
   const attr = attributes || {};
   if (controlData && controlData.name) {
      attr.name = controlData.name;
   } else {
      if (attributes && attributes.name) {
         controlData.name = attributes.name;
      }
   }
   return attr;
}

/**
 * Если существует другой разрешатель имен в config.js. Мы его найдем и используем для подключения.
 * @param tpl
 * @param includedTemplates
 * @param _deps
 * @param config
 * @returns {*}
 */
export function stringTemplateResolver(tpl: string,
                                       includedTemplates: IStringTemplateResolverIncludedTemplates,
                                       _deps: GeneratorEmptyObject,
                                       config: IGeneratorConfig): GeneratorFn {
   const resolver = config && config.resolvers ? Common.findResolverInConfig(tpl, config.resolvers) : undefined;
   if (resolver) {
      return resolver(tpl);
   } else {
      return Common.depsTemplateResolver(tpl, includedTemplates, _deps, config);
   }
}
