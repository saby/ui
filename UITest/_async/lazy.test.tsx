import { renderToString } from 'react-dom/server';
import { act } from 'react-dom/test-utils';
import { Logger } from 'UI/Utils';
import { importer, lazy } from 'UI/Async';

describe('UI/Async:lazy на СП', () => {
    it('Синхронная загрузка контрола на сервере', () => {
        const moduleName = 'UITest/_async/TestControlSync';
        const loggerErrorSpy = jest.spyOn(Logger, 'error').mockImplementation(jest.fn());

        let html;
        act(() => {
            const AsyncLazy = lazy(() => importer(moduleName));
            html = renderToString(<AsyncLazy />);
        });

        expect(loggerErrorSpy).not.toBeCalled();
        expect(html).toMatchSnapshot();
    });

    it('Синхронная загрузка контрола на сервере, с ошибкой', () => {
        const moduleName = 'UITest/_async/Fail/TestControlSync';
        const loggerErrorSpy = jest.spyOn(Logger, 'error').mockImplementation(jest.fn());

        let html;
        act(() => {
            const AsyncLazy = lazy(() => importer(moduleName));
            html = renderToString(<AsyncLazy />);
        });

        expect(loggerErrorSpy).toBeCalledTimes(1);
        expect(loggerErrorSpy.mock.calls[0][0]).toMatchSnapshot('1. ошибка');
        expect(html).toMatchSnapshot('2. верстка');
    });
});
