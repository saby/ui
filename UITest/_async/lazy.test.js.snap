// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/Async:lazy на СП Синхронная загрузка контрола на сервере 1`] = `"<div data-reactroot=\\"\\">TestControlSync</div>"`;

exports[`UI/Async:lazy на СП Синхронная загрузка контрола на сервере, с ошибкой: 1. ошибка 1`] = `"Couldn't load module \\"UITest/_async/Fail/TestControlSync\\""`;

exports[`UI/Async:lazy на СП Синхронная загрузка контрола на сервере, с ошибкой: 2. верстка 1`] = `"<div data-reactroot=\\"\\">У СБИС возникла проблема</div>"`;
